[MLIR][SCF] Sink scf.if from scf.while before region into after region in scf-uplift-while-to-for

NexMing opened the pull request:
When a `scf.if` directly precedes an `scf.condition` in the before region of an `scf.while` and both share the same condition, move the if into the after region of the loop. This helps simplify the control flow to enable uplifting `scf.while` to `scf.for`.

Commit:
[MLIR][SCF] Sink scf.if from scf.while before region into after region.

MaheshRavishankar reviewed (dismissed):
I am not sure this is a canonicalization. This seems like a cost-model based thing. Someone else might want to hoist the if condition

matthias-springer reviewed (commented):
> This preserves semantics while simplifying the condition region and canonicalizing control flow for further optimizations.

What further optimizations did you have in mind?

NexMing commented:
> What further optimizations did you have in mind?

The motivation is that I want to lift `cf` to `scf.for`.
For example, I want to convert the following control flow into an `scf.for`.
```
  func.func @test(%arg0: memref<100xf32>) {
    %c100 = arith.constant 100 : index
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %cst = arith.constant 0.000000e+00 : f32
    cf.br ^bb1(%c0 : index)
  ^bb1(%0: index):  // 2 preds: ^bb0, ^bb2
    %1 = arith.cmpi slt, %0, %c100 : index
    cf.cond_br %1, ^bb2, ^bb3
  ^bb2:  // pred: ^bb1
    memref.store %cst, %arg0[%0] : memref<100xf32>
    %2 = arith.addi %0, %c1 : index
    cf.br ^bb1(%2 : index)
  ^bb3:  // pred: ^bb1
    return
  }
```
When I use the `lift-cf-to-scf` pass, I get:
```
  func.func @test(%arg0: memref<100xf32>) {
    %0 = ub.poison : index
    %c100 = arith.constant 100 : index
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %cst = arith.constant 0.000000e+00 : f32
    %1 = scf.while (%arg1 = %c0) : (index) -> index {
      %2 = arith.cmpi slt, %arg1, %c100 : index
      %3 = scf.if %2 -> (index) {
        memref.store %cst, %arg0[%arg1] : memref<100xf32>
        %4 = arith.addi %arg1, %c1 : index
        scf.yield %4 : index
      } else {
        scf.yield %0 : index
      }
      scf.condition(%2) %3 : index
    } do {
    ^bb0(%arg1: index):
      scf.yield %arg1 : index
    }
    return
  }
```
When I continued using `test-scf-uplift-while-to-for`, the attempt to lift `scf.while` to `scf.for` failed.
After the current optimizations, `test-scf-uplift-while-to-for` can successfully lift `scf.while` to `scf.for`.
I believe the current optimization always simplifies the control flow, so I think it belongs to canonicalization.
I don’t quite understand in what situations there would be a need to hoist the if condition.
If needed, I will consider moving it into `test-scf-uplift-while-to-for`.

NexMing commented:
ping

MaheshRavishankar commented:
> I believe this optimization always simplifies the control flow, so I think it belongs to canonicalization.
I don’t quite understand in what situations there would be a need to hoist the if condition.
If needed, I will consider moving it into test-scf-uplift-while-to-for.

It would be good to not have this as a canonicalization to start with. This seems like a pretty heavy weight transformation to add to canonicalizations.

Commit:
Move the pattern into `populateUpliftWhileToForPattern`.

MaheshRavishankar commented on mlir/lib/Dialect/SCF/Transforms/UpliftWhileToFor.cpp:
```diff
@@ -19,6 +19,137 @@
+    }
+
+    SmallVector<Value> additionalUsedValues;
+    auto isValueUsedInsideIf = [&](Value val) {
```
It might be better to use methods like getValuesDefinedFromAbove, or a backWardSlice to get this information instead of iterating this way.

MaheshRavishankar commented:
> > This seems like a pretty heavy weight transformation to add to canonicalizations.
> 
> Which part? The transformation itself? As long as the matching phase has a "fail fast" path it seems fine to me actually. The most important thing is to do as little work as possible in the pattern when there is nothing to do.

An if-hoisting across a while does not fit in canonicalizations.

Commit:
Use `visitUsedValuesDefinedAbove` to simplify the code.

MaheshRavishankar commented:
> > An if-hoisting across a while does not fit in canonicalizations.
> 
> 1) You're stating it as a fact, but that requires some elaboration: why is it a bad canonicalization?
> 2) Can you please elaborate how your stated opinion here related to it being "heavy weight"? I'm not sure I follow and you're not answering your questions.

I think we need to find a way out of this circular discussion. The onus is to prove that this is canonical. Just stating "I can't think of why you wouldn't do it" isn't justification enough. You can't turn that around and make it an onus on reviewers as to prove why this shouldnt be a canonicalization. Of all possible representations, you need to prove why a canonical representation is always superior.

But in any case, the pattern is doing a whole bunch of code motion and moving instructions around. That is an extremely heavy weight transformation. This is not a targeted "get to a better representation of an operation" transformation.

MaheshRavishankar reviewed (commented):
Oh, I see this was moved back into canonicalization. So I was mistaken on the state of the change. 

Really this is bad practice! I urge you to not add it as a canonicalization, but I dont have the heart to block this change, even though in my book this is following bad practices.

joker-eph commented:
> I’m not opposed to moving it out of canonicalization, but I believe the reason for temporarily moving it out of canonicalization is that this transformation is complex and expensive, which introduces instability and uncertainty., rather than it being unsuitable for canonicalization. Once this optimization becomes sufficiently stable, could we move it back into the canonicalization stage?

If there is "instability and uncertainty", that's a problem to add this in the first place. 
Otherwise I'd like to understand the path to get this into canonicalization? What does it mean for this to "become sufficiently stable"?

Groverkss commented:
<blockquote>
As far as I understand, this transformation is doing:

if (%cond) {
  foo()
}
cond_br %cond ^continue, ^end

^continue:
    ...
^end:
   ...
->

cond_br %cond ^continue, ^end

^continue:
    foo()
    ...
^end:
   ...
</blockquote>

I think this may be over simplifying the transformation a bit. Maybe I'm wrong (so feel free to correct my example if i'm wrong), but let me try to give an example. Let's take the case:

```
while {

%something = some_loop_invariant_compute_but_i_never_hoisted_it_out()
if (%cond) {
  foo(%something)
}
cond_br %cond ^continue, ^end

} while {

^continue:
    ...
^end:
   ...
}
```

If you decide to do the same transformation, you need to somehow also forward `%something`. You need to either:

- Check you don't need to forward anything from the conditional region, which would mean check all uses inside the if block
OR
- Clone the entire backward chain

Both of these seem heavy weight. Note that you cannot really safely do this transformation without doing atleast one of these. Analyze a whole region or cloning an entire backward chain (or finding that backward chain) seems a bit too much for a canonicalization maybe.

Groverkss commented on mlir/lib/Dialect/SCF/Transforms/UpliftWhileToFor.cpp:
What about arguements that are defined inside the while conditional region, but are not addition value arguements?

NexMing replied:
The `visitUsedValuesDefinedAbove` at line 95 will collect all values that are defined outside the `scf.if`.

NexMing commented:
> If there is "instability and uncertainty", that's a problem to add this in the first place. Otherwise I'd like to understand the path to get this into canonicalization? What does it mean for this to "become sufficiently stable"?

What I’m currently worried about is an extreme case where the number of operands to `scf.condition` could grow significantly, and likewise the number of result values of `scf.while` could increase substantially. This might make the `scf.while` look more complex, and I’m not sure whether this would violate the principles of canonicalization.

jpienaar commented:
> What I’m currently worried about is an extreme case where the number of operands to scf.condition could grow significantly, and likewise the number of result values of scf.while could increase substantially. This might make the scf.while look more complex, and I’m not sure whether this would violate the principles of canonicalization.

That is valid concern. I think you are saying appearance to also mean ability to analyze. It could result in more complicated liveness analysis than pre-transform (and to Kumar's point, trade off between recomputing or increasing high water mark).

As a general pattern though that is something that could be analyzed in different flows. It seems desirable/useful pattern but not always on/non canonical, so currently proposed as pattern seems sensible. We don't have a big corpus upstream to measure cost, so I don't think we can put that bar here (of course if folks can find cases where bad and file issues great). It seems useful pattern that can be further refined here.

Commit:
Add more constraints to simplify the transformation process and move it into canonicalization.
We do not have an appropriate cost model for evaluation, and adding more constraints can ensure that it is always beneficial and does not introduce excessive compile-time overhead.

NexMing commented:
Ping!
Adding more restrictions ensures that the transformation is always profitable and does not introduce excessive compile-time overhead.

Commit:
clean up.

NexMing commented on mlir/lib/Dialect/SCF/IR/SCF.cpp:
Because values defined in `before` region cannot reach the `do` region, I need to ensure that the `cond` value is not used inside `scf.if` (although this usually doesn’t happen). Otherwise, I would need to pass or create new values to update the use-chain.

For simplicity, I restrict the `cond` value to have exactly two uses (these two uses come from the `scf.if` and `scf.condition` operations).

joker-eph replied:
> The reason is to avoid creating a new scf.while, which would inflate its result values

I don't quite see the problem with this though, seems like the kind of usual transformations we do.

>  and to avoid additional def-use chain analysis that would increase compile-time overhead.

Do you have examples of compile-time issue? I'd rather be specific and be able to measure the issue here. Collecting values and manipulating def-use chains should be OK here, as long as there is no complex analysis required beyond the direct chains.

> We can treat this as an initial version and improve it gradually later, right?

If you rather go this route, then before merging this I think I'd like to see a chain of stacked/dependent PRs showing the end results with the gradual steps to get there.

NexMing replied:
Sorry for the late reply.

> I don't quite see the problem with this though, seems like the kind of usual transformations we do.

Do you really think adding result values to `scf.while` is fine? From my perspective, it introduces dead values, and they cannot be eliminated by CSE, which makes it risky.
The root cause is that values in the `before` region of `scf.while` cannot reach the `do` region, and passing values to the `do` region is tightly coupled with the result values. This already makes me question whether the design of `scf.while` is reasonable—why does it use two regions instead of two basic blocks?

> Do you have examples of compile-time issue? I'd rather be specific and be able to measure the issue here. Collecting values and manipulating def-use chains should be OK here, as long as there is no complex analysis required beyond the direct chains.

This is just my subjective guess. Maybe I was being lazy, since the revised implementation would require additional analysis and make the code more complex. So I chose to specialize the implementation to meet my current needs. I’ll probably consider improving it further only when new requirements arise.

joker-eph replied:
> Do you really think adding result values to scf.while is fine? From my perspective, it introduces dead values, and they cannot be eliminated by CSE. I can’t accept the idea that introducing undeletable dead values during compilation is risk-free.

These aren't "dead-values" really... They come necessary from the structure of the `scf.while`, if anything this is a deficiency of `scf.while` design.
I don't quite see what CSE has to do with these?
What concrete problem do you expect from this actually?

> This already makes me question whether the design of scf.while is reasonable—why does it use two regions instead of two basic blocks?

I suspect only to be able to keep "single block regions". We could have had a CFG with two terminators: scf.continue and scf.break to differentiate the looping case from the exit case.

By the way, the `scf.loop` I'm introducing here would not have the same issue as `scf.while`: https://github.com/llvm/llvm-project/pull/166688/files#diff-49eb351905b55963bf6871078619f7511e32362d729b2e69f7a59b4b4b204056


If you're interest is to pattern match the specific pattern produced by `lift-cf-to-scf`, then it may be a better fit in this pass than as a canonicalization because of this implicit coupling.

NexMing replied:
> I don't quite see what CSE has to do with these?

Oh, sorry — I mistakenly wrote CSE instead of DCE. I mean that DCE optimization cannot eliminate it.

> seems like the kind of usual transformations we do.

I find it hard to accept the statement. Does this mean that your transformations may also produce dead values that cannot be eliminated? Have you considered extreme cases or stress tests? Or could this actually be a hidden optimization bug that simply hasn’t been discovered yet?

> What concrete problem do you expect from this actually?

Subjectively, I feel that adding result values to `scf.while` may introduce negative optimization effects. Maybe I just don’t understand it well enough — can you explain why it wouldn’t cause any negative optimization?

> If you're interest is to pattern match the specific pattern produced by `lift-cf-to-scf`, then it may be a better fit in this pass than as a canonicalization because of this implicit coupling.

If I can guarantee that it always results in a positive optimization, then I think it’s fine to use it in canonicalization.

joker-eph replied:
> I find it hard to accept the statement. Does this mean that your transformations may also produce dead values that cannot be eliminated? 
>  Have you considered extreme cases or stress tests? Or could this actually be a hidden optimization bug that simply hasn’t been discovered yet?
> Subjectively, I feel that adding result values to scf.while may introduce negative optimization effects. Maybe I just don’t understand it well enough — can you explain why it wouldn’t cause any negative optimization?

I don't see any issue with what you call "dead values" here. But let's not reverse the burden of explanation: if you claim that this is a problem, it's up to you to provide an example (you just need one), not the opposite actually (I can't provide an example for the absence of something in general...).

> If I can guarantee that it always results in a positive optimization, then I think it’s fine to use it in canonicalization.

I was trying to provide you with a way to make progress right now to land your change, I suspect we may not converge on this PR right now.

NexMing commented on mlir/test/Dialect/SCF/canonicalize.mlir:
In this case, the condition at line 3602 will prevent it.
```
    if (!llvm::equal(ifOp->getResults(), term.getArgs()) ||
        !llvm::equal(ifOp.elseYield()->getOperands(), op.getBeforeArguments()))
      return failure();
```

Commit:
Update the comments.

Commit:
Move pattern into `populateUpliftWhileToForPatterns`.

joker-eph reviewed (approved):
LGTM
